### PR TITLE
feat(ui): Add banner to broadcast messages

### PR DESCRIPTION
**Pull request by rudra-superrr**
Signed-off-by: rchopra <rudra.chopra@siemens.com>

Issue:  Add banner to broadcast messages (#1595)

Description : This functionality can be edited by Admins only.

### How To Test?

1) Click on 9 dots on the top right section of the page
2) Go to control panel -> click custom fields under configuration
3) Select User under Resource section
4) There you will see BannerMessage custom field.
5) Click on three dots to edit this field. Here you can enter message that you want to be displayed.
6) Right click on three dots and give view permission to all sw360 users/roles and change other permissions too. (Refer the ss for this step)

### Screenshots

controlPanel
Click custom field under control panel section

User
Select User under resource section

bannerMessage
Banner Message custom field should look like this

Permissions
Give view permission to all sw360users/roles and make changes in other permissions too

welcomePage
Banner message on top of the page

### CODE REVIEW

**Review by akapti — CHANGES_REQUESTED** [review]
Suggested one minor change.
Failed to test the PR because of the following error:
```
Caused by: freemarker.core._TemplateModelException: Java method "com.liferay.portlet.expando.model.impl.ExpandoBridgeImpl.getAttribute(String)" threw an exception when invoked on com.liferay.portlet.expando.model.impl.ExpandoBridgeImpl object "com.liferay.portlet.expando.model.impl.ExpandoBridgeImpl@4233b486"; see cause exception in the Java stack trace.__----_FTL stack trace ("~" means nesting-related):_- Failed at: #assign expandoAttribute = liferay_us...  [in template "org.eclipse.sw360.liferay-theme_SERVLET_CONTEXT_/templates/portal_normal.ftl" at line 42, column 1]_---- [Sanitized]
```

It might be because by default the banner message field is not set in Liferay custom fields section. You should create the and set the default value if the custom field is not available in Liferay on startup.

File: frontend/liferay-theme/src/main/webapp/js/banner.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+function clickCounter() {
+	sessionStorage.setItem('header', true);
+	document.getElementById('updateMessage').style.display = 'none';
+}
+
+function second() {
+	if (sessionStorage.getItem('header') == 'true') {
+		document.getElementById('updateMessage').style.display = 'none';
+	}
+
```
[the remark this answers is not in the document]
Comment (rudra-superrr) [reply frontend/liferay-theme/src/main/webapp/js/banner.js:20]: You need to set the custom field value manually before pulling the code.
Please refer the above steps for testing.

File: frontend/liferay-theme/src/main/webapp/templates/portal_normal.ftl [hunk 2]
```diff
@@ -59,6 +66,12 @@
 <div id="wrapper" class="${hide_portlet_edit_decorators_css} ${login_css}">
 	<header id="banner" role="banner">
 		<div id="heading" class="container">
+            <#if expandoAttribute?has_content>
+                <div id="updateMessage">
+                    <span class="closebtn" onclick="clickCounter()">&times;</span>
+                    <strong>Server Update!</strong> Indicates ${expandoAttribute}.
```
Comment (akapti) [inline frontend/liferay-theme/src/main/webapp/templates/portal_normal.ftl:72]: I am not sure if this default text `<strong>Server Update!</strong> Indicates` is really required.

File: frontend/liferay-theme/src/main/webapp/css/portal/_header.scss [hunk 3]
```diff
@@ -24,6 +24,27 @@
 		}
 	}
 
+    #updateMessage {
+        padding: 20px;
+        background-color: #F4A460;
+        color: white;
+        }
```
Comment (akapti) [inline frontend/liferay-theme/src/main/webapp/css/portal/_header.scss:31]: fix the indentation.. also on line `42` and `46` .

File: frontend/liferay-theme/src/main/webapp/templates/portal_normal.ftl [hunk 4]
```diff
@@ -59,6 +65,15 @@
 <div id="wrapper" class="${hide_portlet_edit_decorators_css} ${login_css}">
 	<header id="banner" role="banner">
 		<div id="heading" class="container">
+            <#if is_signed_in>
+                <#assign expandoAttribute = liferay_user.getExpandoBridge().getAttribute("BannerMessage") />
+                <#if expandoAttribute?has_content>
+                    <div id="updateMessage">
+                        <span class="closebtn" onclick="clickCounter()">&times;</span>
```
Comment (akapti) [inline frontend/liferay-theme/src/main/webapp/templates/portal_normal.ftl:72]: please rename the function `clickCounter()` to more meaningful (based on action being performed on click of UI element)

**Review by akapti — CHANGES_REQUESTED** [review]
Suggested some minor review comments.
Found one issue while testing, not able to close the banner message.

image

**Comment by rudra-superrr** [comment]
Hi @akapti, done the changes.

File: frontend/liferay-theme/src/main/webapp/templates/portal_normal.ftl (hunk 4 above)
Comment (rudra-superrr) [reply frontend/liferay-theme/src/main/webapp/templates/portal_normal.ftl]: my bad, forgot to change it to closeBanner.

**Review by akapti — APPROVED** [review]
Tested the changes, working as expected.
FYI: The banner message will be visible again in new browser tab (tab-y) even after closing it in old tab (tab-x).

Code looks good.

**Pull request merged**